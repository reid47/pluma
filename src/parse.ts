import { ParseError } from './errors';
import * as t from './types';
import * as nodes from './nodes';

class Parser {
  private readonly tokens: t.Token[];
  private readonly source: string;
  private readonly tokenCount: number;
  private index: number;
  private token: t.Token;
  private eof: boolean;

  constructor(tokens: t.Token[], source: string) {
    this.tokens = tokens;
    this.source = source;
    this.tokenCount = this.tokens.length;
    this.index = 0;
    this.token = this.tokens[this.index];
    this.eof = false;
  }

  private advance(amount: number = 1) {
    this.index += amount;
    this.token = this.tokens[this.index];
    if (this.index >= this.tokenCount) this.eof = true;
  }

  private tokenIs(kind: t.TokenKind, nextKind?: t.TokenKind) {
    if (this.token.kind !== kind) return false;

    if (nextKind) {
      const nextToken = this.tokens[this.index + 1];
      if (!nextToken) return false;
      if (nextToken.kind !== nextKind) return false;
    }

    return true;
  }

  private parseComments(): void {
    while (!this.eof && this.tokenIs('Comment')) {
      this.advance();
    }
  }

  private parseParenthetical(): t.Expression {
    if (!this.tokenIs('LeftParen')) return;

    const { lineStart, colStart } = this.token;

    this.advance();

    const expr = this.parseExpression();

    if (!this.tokenIs('RightParen')) {
      throw new ParseError(
        this.token.lineStart,
        this.token.lineEnd,
        `Expected a ) to match ( at line ${lineStart}, column ${colStart}, but found ${
          this.token.kind
        }`
      );
    }

    this.advance();

    return expr;
  }

  private parseAssignmentExpression(): t.AssignmentExpression {
    if (!this.tokenIs('Identifier', 'Equals')) return;

    const id = this.parseIdentifier();

    if (/^[A-Z]/.test(id.name)) {
      throw new ParseError(
        id.lineStart,
        id.colStart,
        `Cannot assign to type name '${
          id.name
        }'. If you meant to declare a type, use the 'type' keyword. ` +
          `If you meant to assign to a variable, it must start with a lowercase letter.`
      );
    }

    this.advance();

    const expr = this.parseExpression();

    return new nodes.AssignmentExpression(
      id,
      expr,
      id.lineStart,
      id.colStart,
      expr.lineEnd,
      expr.colEnd
    );
  }

  private parseIdentifier(): t.Identifier {
    if (!this.tokenIs('Identifier')) return;

    const { value, lineStart, lineEnd, colStart, colEnd } = this.token;

    this.advance();

    return new nodes.Identifier(value, lineStart, colStart, lineEnd, colEnd);
  }

  private parseBlock(): t.Block {
    if (!this.tokenIs('LeftBrace')) return;

    const { lineStart, colStart } = this.token;

    this.advance();

    const parameters: t.Identifier[] = [];
    const firstExpression = this.parseExpression();

    let body;

    if (
      firstExpression.kind === 'Identifier' &&
      (this.tokenIs('Comma') || this.tokenIs('DoubleArrow'))
    ) {
      parameters.push(firstExpression);

      while (this.tokenIs('Comma')) {
        this.advance();

        const parameter = this.parseIdentifier();

        if (!parameter) {
          throw new ParseError(-1, -1, 'Expected a parameter name after , in block');
        }

        parameters.push(parameter);
      }

      if (!this.tokenIs('DoubleArrow')) {
        throw new ParseError(
          this.token.lineStart,
          this.token.lineEnd,
          `Expected a => after block parameters, but found ${this.token.kind}`
        );
      }

      this.advance();

      body = this.parseExpression();

      if (!body) {
        throw new ParseError(-1, -1, 'Expected an expression after => in block');
      }
    } else {
      body = firstExpression;
    }

    if (!this.tokenIs('RightBrace')) {
      throw new ParseError(
        this.token.lineStart,
        this.token.lineEnd,
        `Expected a } to close block, but found ${this.token.kind}`
      );
    }

    const { lineEnd, colEnd } = this.token;

    this.advance();

    return new nodes.Block(parameters, body, lineStart, colStart, lineEnd, colEnd);
  }

  private parseBooleanLiteral(): t.BooleanLiteral {
    if (!this.tokenIs('Boolean')) return;

    const { value, lineStart, lineEnd, colStart, colEnd } = this.token;

    this.advance();

    return new nodes.BooleanLiteral(
      value === 'true',
      lineStart,
      colStart,
      lineEnd,
      colEnd
    );
  }

  private parseNumericLiteral(): t.NumericLiteral {
    if (!this.tokenIs('Number')) return;

    const { value, lineStart, lineEnd, colStart, colEnd } = this.token;
    const style = value.indexOf('.') > -1 ? 'float' : 'integer';
    const numericValue = Number(value);

    this.advance();

    return new nodes.NumericLiteral(
      style,
      numericValue,
      value,
      lineStart,
      colStart,
      lineEnd,
      colEnd
    );
  }

  parseStringLiteral(): t.StringLiteral {
    if (!this.tokenIs('String')) return;

    const { value, lineStart, lineEnd, colStart, colEnd } = this.token;

    this.advance();

    return new nodes.StringLiteral(value, lineStart, colStart, lineEnd, colEnd);
  }

  parseExpression(): t.Expression {
    if (this.eof) return;

    this.parseComments();

    if (!this.token) return;

    return (
      this.parseParenthetical() ||
      this.parseAssignmentExpression() ||
      this.parseBlock() ||
      this.parseIdentifier() ||
      this.parseBooleanLiteral() ||
      this.parseNumericLiteral() ||
      this.parseStringLiteral()
    );
  }

  parseModule(): t.Module {
    const definitions: t.Definition[] = [];
    const body: t.Expression[] = [];

    let node;

    while ((node = this.parseExpression())) {
      body.push(node);
    }

    const bodyLineEnd = body.length ? body[body.length - 1].lineEnd : 0;
    const bodyColEnd = body.length ? body[body.length - 1].colEnd : 0;

    return new nodes.Module(definitions, body, 1, 0, bodyLineEnd, bodyColEnd);
  }
}

export function parseExpression(tokens: t.Token[], source: string): t.Expression {
  return new Parser(tokens, source).parseExpression();
}

export function parseModule(tokens: t.Token[], source: string): t.Module {
  return new Parser(tokens, source).parseModule();
}
