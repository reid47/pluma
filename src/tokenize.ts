import * as t from './tokens';
import { ParseError } from './errors';
export { tokenize };

function tokenize(source: string): t.Token[] {
  const tok = new Tokenizer(source);

  let token;
  while ((token = readToken(tok))) {
    if (Array.isArray(token)) tok.tokens.push(...token);
    else tok.tokens.push(token);
  }

  return tok.tokens;
}

class Tokenizer {
  readonly source: string;
  readonly chars: string[];
  readonly length: number;
  readonly tokens: t.Token[];
  index: number;
  char: string;
  line: number;
  lineStartIndex: number;
  eof: boolean;

  constructor(source: string) {
    this.source = source;
    this.chars = Array.from(this.source);
    this.length = this.chars.length;
    this.tokens = [];
    this.char = this.chars[0];
    this.index = 0;
    this.line = 1;
    this.lineStartIndex = 0;
    this.eof = false;
  }
}

function advance(tok: Tokenizer, amount = 1) {
  tok.index += amount;
  tok.char = tok.chars[tok.index];
  if (tok.index >= tok.length) tok.eof = true;
}

function charIs(tok: Tokenizer, char: string, nextChar?: string, nextNextChar?: string) {
  if (tok.char !== char) return false;
  if (nextChar && tok.chars[tok.index + 1] !== nextChar) return false;
  if (nextNextChar && tok.chars[tok.index + 2] !== nextChar) return false;
  return true;
}

function prevCharIs(tok: Tokenizer, testChar: string) {
  return tok.chars[tok.index - 1] === testChar;
}

function readComment(tok: Tokenizer): t.Token | void {
  if (!charIs(tok, '#')) return;

  const colStart = tok.index - tok.lineStartIndex;
  let value = '';
  let skip = true;

  while (!charIs(tok, '\n') && !tok.eof) {
    if (!skip) value += tok.char;
    advance(tok);
    skip = false;
  }

  const colEnd = tok.index - tok.lineStartIndex;

  return {
    kind: 'Comment',
    value,
    lineStart: tok.line,
    lineEnd: tok.line,
    colStart,
    colEnd
  };
}

function readIdentifier(tok: Tokenizer): t.Token | void {
  if (!isIdentifierStartChar(tok.char)) return;

  let value = '';
  const colStart = tok.index - tok.lineStartIndex;

  while (isIdentifierChar(tok.char) && !tok.eof) {
    value += tok.char;
    advance(tok);
  }

  const colEnd = tok.index - tok.lineStartIndex;

  if (value === 'true' || value === 'false') {
    return {
      kind: 'Boolean',
      value,
      lineStart: tok.line,
      lineEnd: tok.line,
      colStart,
      colEnd
    };
  }

  return {
    kind: 'Identifier',
    value,
    lineStart: tok.line,
    lineEnd: tok.line,
    colStart,
    colEnd
  };
}

function readChar(tok: Tokenizer): t.Token | void {
  if (!charIs(tok, "'")) return;

  const colStart = tok.index - tok.lineStartIndex;

  advance(tok);

  const value = tok.char;

  advance(tok);

  if (!charIs(tok, "'")) {
    throw new ParseError(
      tok.line,
      tok.index - tok.lineStartIndex,
      "Expected closing ' after character"
    );
  }

  const colEnd = tok.index - tok.lineStartIndex;

  advance(tok);

  return {
    kind: 'Char',
    value,
    lineStart: tok.line,
    lineEnd: tok.line,
    colStart,
    colEnd
  };
}

function readNumber(tok: Tokenizer): t.Token | void {
  if (!isDecimalDigit(tok.char)) return;

  const colStart = tok.index - tok.lineStartIndex;

  let value = '';

  if (charIs(tok, '0', 'x')) {
    value += '0x';
    advance(tok, 2);

    while (isHexDigit(tok.char) && !tok.eof) {
      value += tok.char;
      advance(tok);
    }

    const colEnd = tok.index - tok.lineStartIndex;

    return {
      kind: 'Number',
      value,
      lineStart: tok.line,
      lineEnd: tok.line,
      colStart,
      colEnd
    };
  }

  if (charIs(tok, '0', 'o')) {
    value += '0o';
    advance(tok, 2);

    while (isOctalDigit(tok.char) && !tok.eof) {
      value += tok.char;
      advance(tok);
    }

    const colEnd = tok.index - tok.lineStartIndex;

    return {
      kind: 'Number',
      value,
      lineStart: tok.line,
      lineEnd: tok.line,
      colStart,
      colEnd
    };
  }

  if (charIs(tok, '0', 'b')) {
    value += '0b';
    advance(tok, 2);

    while (isBinaryDigit(tok.char) && !tok.eof) {
      value += tok.char;
      advance(tok);
    }

    const colEnd = tok.index - tok.lineStartIndex;

    return {
      kind: 'Number',
      value,
      lineStart: tok.line,
      lineEnd: tok.line,
      colStart,
      colEnd
    };
  }

  while (isDecimalDigit(tok.char) && !tok.eof) {
    value += tok.char;
    advance(tok);
  }

  if (charIs(tok, '.')) {
    value += tok.char;
    advance(tok);

    while (isDecimalDigit(tok.char) && !tok.eof) {
      value += tok.char;
      advance(tok);
    }
  }

  if (charIs(tok, 'e')) {
    value += tok.char;
    advance(tok);

    while (isDecimalDigit(tok.char) && !tok.eof) {
      value += tok.char;
      advance(tok);
    }
  }

  const colEnd = tok.index - tok.lineStartIndex;

  return {
    kind: 'Number',
    value,
    lineStart: tok.line,
    lineEnd: tok.line,
    colStart,
    colEnd
  };
}

function readString(tok: Tokenizer): t.Token[] | void {
  if (!charIs(tok, '"')) return;

  const tripleQuoted = charIs(tok, '"', '"', '"');
  const colStart = tok.index - tok.lineStartIndex;
  const quoteSize = tripleQuoted ? 3 : 1;
  const stringTokens: t.Token[] = [];

  let lineStart = tok.line;
  let value = '';

  advance(tok, quoteSize);

  while (!tok.eof) {
    if (charIs(tok, '\n')) {
      tok.line++;
      tok.lineStartIndex = tok.index + 1;
    }

    if (!tripleQuoted) {
      if (charIs(tok, '"') && !prevCharIs(tok, '\\')) break;
    } else {
      if (charIs(tok, '"', '"', '"')) break;
    }

    if (charIs(tok, '$', '(') && !prevCharIs(tok, '\\')) {
      const col = tok.index - tok.lineStartIndex;

      advance(tok, 2);

      stringTokens.push({
        kind: 'String',
        value,
        lineStart,
        lineEnd: tok.line,
        colStart,
        colEnd: col
      });

      stringTokens.push({
        kind: 'InterpolationStart',
        value: '$(',
        lineStart: tok.line,
        lineEnd: tok.line,
        colStart: col,
        colEnd: col + 2
      });

      value = '';

      const parenStack = [];

      let innerToken;

      while ((innerToken = readToken(tok))) {
        if (Array.isArray(innerToken)) {
          stringTokens.push(...innerToken);
          continue;
        }

        if (innerToken.kind === '(') {
          parenStack.push(true);
        } else if (innerToken.kind === ')') {
          if (!parenStack.length) {
            stringTokens.push({
              kind: 'InterpolationEnd',
              value: ')',
              lineStart: tok.line,
              lineEnd: tok.line,
              colStart: innerToken.colStart,
              colEnd: innerToken.colEnd
            });

            lineStart = tok.line;

            break;
          }

          parenStack.pop();
        }

        stringTokens.push(innerToken);
      }
    } else {
      value += tok.char;
      advance(tok);
    }
  }

  if (!tripleQuoted && !charIs(tok, '"')) {
    const col = tok.index - tok.lineStartIndex;

    throw new ParseError(
      tok.line,
      col,
      `Missing a closing " for string starting on line ${lineStart}`
    );
  }

  if (tripleQuoted && !charIs(tok, '"', '"', '"')) {
    const col = tok.index - tok.lineStartIndex;

    throw new ParseError(
      tok.line,
      col,
      `Missing a closing """ for string starting on line ${lineStart}`
    );
  }

  advance(tok, quoteSize);

  const colEnd = tok.index - tok.lineStartIndex;

  stringTokens.push({
    kind: 'String',
    value,
    lineStart,
    lineEnd: tok.line,
    colStart,
    colEnd
  });

  return stringTokens;
}

function readSymbol(
  tok: Tokenizer,
  kind: t.TokenKind,
  firstChar: string,
  secondChar?: string
): t.Token | void {
  if (!charIs(tok, firstChar, secondChar)) return;

  const size = secondChar ? 2 : 1;
  const colStart = tok.index - tok.lineStartIndex;

  advance(tok, size);

  return {
    kind,
    value: kind,
    lineStart: tok.line,
    lineEnd: tok.line,
    colStart,
    colEnd: colStart + size
  };
}

function readOperator(tok: Tokenizer): t.Token | void {
  if (!isOperatorChar(tok.char)) return;

  const colStart = tok.index - tok.lineStartIndex;

  let value = '';

  while (isOperatorChar(tok.char)) {
    value += tok.char;
    advance(tok);
  }

  const colEnd = tok.index - tok.lineStartIndex;

  return {
    kind: 'Operator',
    value,
    lineStart: tok.line,
    lineEnd: tok.line,
    colStart,
    colEnd
  };
}

function readWhitespace(tok: Tokenizer) {
  while (isWhitespace(tok.char) && !tok.eof) {
    if (charIs(tok, '\n')) {
      tok.line++;
      tok.lineStartIndex = tok.index + 1;
    }

    advance(tok);
  }
}

function readToken(tok: Tokenizer): t.Token | t.Token[] | void {
  if (tok.eof) return;

  readWhitespace(tok);

  return (
    readIdentifier(tok) ||
    readNumber(tok) ||
    readChar(tok) ||
    readString(tok) ||
    readSymbol(tok, '->', '-', '>') ||
    readSymbol(tok, ':=', ':', '=') ||
    readSymbol(tok, '=>', '=', '>') ||
    readSymbol(tok, '{', '{') ||
    readSymbol(tok, '}', '}') ||
    readSymbol(tok, '(', '(') ||
    readSymbol(tok, ')', ')') ||
    readSymbol(tok, '[', '[') ||
    readSymbol(tok, ']', ']') ||
    readSymbol(tok, '.', '.') ||
    readSymbol(tok, ',', ',') ||
    readSymbol(tok, ':', ':') ||
    readSymbol(tok, '=', '=') ||
    readOperator(tok) ||
    readComment(tok)
  );
}

function isWhitespace(char: string) {
  switch (char) {
    case ' ':
    case '\n':
    case '\t':
    case '\r':
      return true;
    default:
      return false;
  }
}

function isIdentifierStartChar(char: string) {
  if (!char) return false;

  const codePoint = char.codePointAt(0) || -1;

  return (
    (codePoint >= 97 && codePoint <= 122) ||
    (codePoint >= 65 && codePoint <= 90) ||
    codePoint === 95
  );
}

function isIdentifierChar(char: string) {
  return isIdentifierStartChar(char) || isDecimalDigit(char);
}

function isDecimalDigit(char: string) {
  switch (char) {
    case '0':
    case '1':
    case '2':
    case '3':
    case '4':
    case '5':
    case '6':
    case '7':
    case '8':
    case '9':
      return true;
    default:
      return false;
  }
}

function isHexDigit(char: string) {
  return (
    isDecimalDigit(char) || (char >= 'a' && char <= 'f') || (char >= 'A' && char <= 'F')
  );
}

function isOctalDigit(char: string) {
  switch (char) {
    case '0':
    case '1':
    case '2':
    case '3':
    case '4':
    case '5':
    case '6':
    case '7':
      return true;
    default:
      return false;
  }
}

function isBinaryDigit(char: string) {
  switch (char) {
    case '0':
    case '1':
      return true;
    default:
      return false;
  }
}

function isOperatorChar(char: string) {
  switch (char) {
    case '@':
    case '<':
    case '>':
    case '=':
    case '|':
    case '!':
    case '+':
    case '*':
    case '-':
      return true;
    default:
      return false;
  }
}
