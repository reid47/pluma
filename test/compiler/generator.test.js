import link from '../../src/compiler/linker';
import tokenize from '../../src/compiler/tokenizer';
import parse from '../../src/compiler/parser';
import generate from '../../src/compiler/generator';
import path from 'path';

const sourceDirectory = path.resolve(__dirname, './fixtures');

function expectJsFromSource(source) {
  const tokens = tokenize({ source });
  const ast = parse({ source, tokens });

  return expectJs([ast]);
}

function expectJsFromFile(entryFile) {
  const { asts, entryExports } = link({
    entry: path.resolve(sourceDirectory, entryFile)
  });

  return expectJs(asts, entryExports);
}

function expectJs(asts, entryExports) {
  const js = generate({ asts, entryExports });
  expect(js).toBeDefined();

  let evalError;
  try {
    eval(js);
  } catch (err) {
    evalError = err;
    console.log(js); //eslint-disable-line
    throw err;
  }
  expect(evalError).not.toBeDefined();

  return js;
}

describe('generate', () => {
  test('empty module', () => {
    const js = expectJsFromSource('module TestModule');
    expect(js).toBe(`(function(global, factory) {
  typeof exports === 'object' && typeof module !== 'undefined' ? module.exports = factory() :
  typeof define === 'function' && define.amd ? define(factory) :
  (global.Pluma = factory());
})(this, function() { 'use strict';

var module$TestModule = (function() {
    return {};
})();
});`);
  });

  test('module with exports from entry point', () => {
    const js = expectJsFromFile('Exporting.plum');
    expect(js).toContain('return module$Exporting;');
  });

  test('lots of simple examples', () => {
    expectJsFromFile('HelloWorld.plum');
  });
});
