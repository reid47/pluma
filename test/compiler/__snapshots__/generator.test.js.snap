// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generate empty module 1`] = `
"(function(global, factory) {
  typeof exports === 'object' && typeof module !== 'undefined' ? module.exports = factory() :
  typeof define === 'function' && define.amd ? define(factory) :
  (global.Pluma = factory());
})(this, function() { 'use strict';

var module$TestModule = (function() {
    return {};
})();
});"
`;

exports[`generate lots of simple examples 1`] = `
"(function(global, factory) {
  typeof exports === 'object' && typeof module !== 'undefined' ? module.exports = factory() :
  typeof define === 'function' && define.amd ? define(factory) :
  (global.Pluma = factory());
})(this, function() { 'use strict';

var module$HelloWorld = (function() {
  var b, list, fn, conditional, greet, withLetExpression, obj, p1, p2, p3, piped, piped2;

  b = 47;

  list = [1, 2, 3, 4];

  fn = function(a) {
    return function(b) {
      return function(c) {
        return \\"hello, world!\\";
      };
    };
  };

  conditional = function(x) {
    return (x ? true : false);
  };

  greet = function(name) {
    return (\\"hi \\" + name + \\"!\\");
  };

  withLetExpression = function(firstName) {
    return (function() {
      var lastName;
      lastName = \\"test\\";

      return (\\"hi \\" + firstName + \\" \\" + lastName + \\"!\\");
    })();
  };

fn(1)(2)(3);

  obj = { a: 1, b: b };

  p1 = function(s) {
    return \\"nice\\";
  };

  p2 = function(s) {
    return \\"ok\\";
  };

  p3 = function(a) {
    return function(b) {
      return \\"test\\";
    };
  };

  piped = greet(p3(\\"hi\\")(p1(\\"reid\\")));

  piped2 = greet(p1(\\"hello\\"));

  return { b: b, fn: fn };
})();

return module$HelloWorld;
});"
`;

exports[`generate module with exports from entry point 1`] = `
"(function(global, factory) {
  typeof exports === 'object' && typeof module !== 'undefined' ? module.exports = factory() :
  typeof define === 'function' && define.amd ? define(factory) :
  (global.Pluma = factory());
})(this, function() { 'use strict';

var module$Exporting = (function() {
  var func1, func2;

  func1 = function(x) {
    return 47;
  };

  func2 = function(y) {
    return \\"hi\\";
  };

  return { func1: func1, func2: func2 };
})();

return module$Exporting;
});"
`;
